### ISSUE TITLE
new approach: Limitations on `Component::Root`

### ISSUE THREAD

**Issue by cofee-on-the-desk** [Relm4/Relm4#143]
`Root` requires a type to be `OnDestroy`, which is currently only implemented for gtk widgets. This makes it impossible to make [`gtk::FileChooserNative`](https://gtk-rs.org/gtk4-rs/stable/latest/docs/gtk4/struct.FileChooserNative.html) the component root. If we could fix this, it would benefit #141 and we won't have to use non-native dialogs.

**Comment by mmstick** [comment]
What makes it impossible to implement `OnDestroy`?

**Comment by AaronErhardt** [comment]
Native dialogs are no regular widgets and don't implement `WidgetExt` which makes it impossible to use [connect_destroy](https://gtk-rs.org/gtk4-rs/stable/latest/docs/gtk4/prelude/trait.WidgetExt.html#tymethod.connect_destroy). Also, I couldn't find another mechanism that'd provide something similar for native dialogs.

I'm wondering if we can remove this requirement and rather bind the lifetime of the internal runtime to the lifetime of the controller.

**Comment by mmstick** [comment]
You can use [connect_response](https://gtk-rs.org/gtk4-rs/stable/latest/docs/gtk4/prelude/trait.NativeDialogExt.html#tymethod.connect_response) and listen for `ResponseType::DeleteEvent`.

**Comment by AaronErhardt** [comment]
The idea is good, but we still can't have a generic implementation over all widgets that are `AsRef<NativeDialog>` because they could theoretically be `AsRef<Widget>`, too and Rust doesn't like that.

We could do it manually or with a macro for each native dialog type though...

**Comment by mmstick** [comment]
This may work https://crates.io/crates/syllogism

**Comment by AaronErhardt** [comment]
I wish specialization was stable already, but I don't think using this experimental crate is a good solution. 

The question is, what would be the harm in storing a type inside the component structures that drops the runtime when itself is dropped. It would be nice to have the runtime bound to the root widget but on the other side it might be more predictable to use the `Controller` to manage the lifetime of the update loop.

Or we can use the lifetime of this closure and move in an object that drops the runtime when it's dropped: https://gtk-rs.org/gtk-rs-core/git/docs/glib/object/trait.ObjectExt.html#tymethod.watch_closure
Haven't used that API yet, but if we make it happen I guess it will be through a hack like this.

**Comment by mmstick** [comment]
I could attach a sender to `ComponentSender` which requests to stop the event loop, and then the author of the component could manually set up the kill operation of their choice

**Comment by mmstick** [comment]
Something like this:

```rs
impl ComponentSenderInner {
    /// Sends a kill signal which will stop this component's event loop.
    pub fn kill(&self) {
        let _ = self.killswitch.send(());
    }

    /// Attaches the kill signal to a widget.
    pub fn kill_with_widget(self: &Rc<Self>, widget: &dyn AsRef<gtk::Widget>) {
        use gtk::prelude::WidgetExt;
        let this = self.clone();
        widget.as_ref().connect_destroy(move |_| this.kill());
    }

    /// Attahes the kill signal to a native dialog
    pub fn kill_with_native_dialog(self: &Rc<Self>, dialog: &dyn AsRef<gtk::NativeDialog>) {
        use gtk::prelude::NativeDialogExt;
        let this = self.clone();
        dialog.as_ref().connect_response(move |_, response| {
            if let gtk::ResponseType::DeleteEvent = response {
                this.kill();
            }
        });
    }
}
```

Then setting up like so:

```rs
fn init(
    counter: Self::InitParams,
    root: &Self::Root,
    sender: &ComponentSender<Self>,
) -> ComponentParts<Self> {
    let model = AppModel { counter };

    sender.kill_with_widget(root);

    // Insert the macro codegen here
    let widgets = view_output!();

    ComponentParts { model, widgets }
}
```
```

**Comment by AaronErhardt** [comment]
Closed via #179 (you might have to use `cargo update` to get the latest gtk-rs commits)
